Removing C++11-isms about multithreading.

nicolasnoble opened the pull request:
This is more a request for comments about the general idea of the code rather than its coding style. This doesn't work yet, and there'd be a few more things needed to turn it into a real commit. I probably want to split sync.h up into thd.h. I also need to add more comments. And in general, I guess we want to use capitalized names for the class names.

But before I spend more time on this, I'd like to gather comments from people if this is about the right direction we'd go for this.

Reminder: Visual Studio 2010 doesn't support `#include <mutex>`, `#include <condition_variable>` and `#include <thread>`. So this is an attempt at cleaning that up a bit.

ctiller commented:
I would probably lean towards using the C++11 names if possible - but in a
name space defined by is. 2010 is not going to matter at all in a year or
two, but our codebase will grow in that time - and it will be gratifying to
just need a simple set of cls to clean everything up.

On Thu, Apr 2, 2015, 6:21 PM Nicolas Noble <notifications@github.com> wrote:

> This is more a request for comments about the general idea of the code
> rather than its coding style. This doesn't work yet, and there'd be a few
> more things needed to turn it into a real commit. I probably want to split
> sync.h up into thd.h. I also need to add more comments. And in general, I
> guess we want to use capitalized names for the class names.
>
> But before I spend more time on this, I'd like to gather comments from
> people if this is about the right direction we'd go for this.
>
> Reminder: Visual Studio 2010 doesn't support #include , #include and
> #include . So this is an attempt at cleaning that up a bit.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/grpc/grpc/pull/1185
> Commit Summary
>
>    - Starting off by removing all offending headers.
>    - Rewriting mutex, condition_variable, and thread.
>
> File Changes
>
>    - *A* include/grpc++/impl/sync.h
>    <https://github.com/grpc/grpc/pull/1185/files#diff-0> (126)
>    - *M* include/grpc++/server.h
>    <https://github.com/grpc/grpc/pull/1185/files#diff-1> (7)
>    - *M* include/grpc/support/thd.h
>    <https://github.com/grpc/grpc/pull/1185/files#diff-2> (2)
>    - *M* src/cpp/server/server.cc
>    <https://github.com/grpc/grpc/pull/1185/files#diff-3> (24)
>    - *M* src/cpp/server/server_context.cc
>    <https://github.com/grpc/grpc/pull/1185/files#diff-4> (15)
>    - *M* src/cpp/server/thread_pool.cc
>    <https://github.com/grpc/grpc/pull/1185/files#diff-5> (16)
>    - *M* src/cpp/server/thread_pool.h
>    <https://github.com/grpc/grpc/pull/1185/files#diff-6> (11)
>
> Patch Links:
>
>    - https://github.com/grpc/grpc/pull/1185.patch
>    - https://github.com/grpc/grpc/pull/1185.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/1185>.
>

sunwangme commented:
plz use c++11.  http2 + protobuf3 + c++11.

nicolasnoble commented:
@sunwangme: we can't, if we want to support older compilers, such as Visual Studio 2010 - and we do want to support it.
@ctiller: sure - that's actually helping for the cleanup phase I'd do next on that PR.

nicolasnoble commented:
Alright, cleaned up and updated. This should no longer break the current build, as it now piggy backs on C++11 if available. But it'll only work under Visual Studio 2010 if we get #1191 in.

sunwangme commented:
thanks

nicolasnoble commented:
So I believe this PR is now in a state that is polished enough to be submitted for an actual review.

vjpai commented on include/grpc++/impl/sync.h:
```diff
@@ -0,0 +1,46 @@
+#include <grpc++/impl/sync_nocxx11.h>
+#else
+#include <grpc++/impl/sync_cxx11.h>
+
```
Why the blank line here?

nicolasnoble replied:
Probably PEBCAK. Fixed.

vjpai commented on include/grpc++/impl/sync_no_cxx11.h:
```diff
@@ -0,0 +1,98 @@
+
+template <class mutex>
+class lock_guard {
+ public:
```
So, this has an API different from the C++ standard. I would say that what you have as lock_guard should be unique_lock, and then lock_guard should be unique_lock with the lock and unlock methods deleted.

nicolasnoble replied:
So, how about the new change I'm proposing. I still have the inheritance of unique_lock over lock_guard, but that's really to try to maintain a single codepath for basically doing the same thing. I hate having two different codepaths for doing the same thing: it just calls for errors when fixing a problem with one, and forgetting the other one.

This time however I made lock / unlock be protected, and made available from unique_lock. This way we get the lock / unlock semantic available only in the unique_lock version.

vjpai commented on include/grpc++/impl/sync_no_cxx11.h:
```diff
@@ -0,0 +1,98 @@
+ public:
+   unique_lock(mutex &mu) : lock_guard(mu) { }
+};
+
```
Similar to my above comment. I am concerned about creating inheritance relationships in the no-C++11 build that aren't present in the actual C++11 build. Composition seems like a safer path than inheritance to reduce the likelihood that some programmer somewhere actually confounds the two in a piece of code that we have to sort out later.

vjpai commented on include/grpc++/impl/sync_no_cxx11.h:
```diff
@@ -0,0 +1,98 @@
+  condition_variable() { gpr_cv_init(&cv_); }
+  ~condition_variable() { gpr_cv_destroy(&cv_); }
+  void wait(lock_guard<mutex> &mu) {
+    mu.locked = false;
```
So with my above comments in mind, this should be a unique lock rather than a lock guard..... matching the C++11 standard again.

vjpai commented on include/grpc++/impl/thd_no_cxx11.h:
```diff
@@ -0,0 +1,81 @@
+    func_ = new thread_function<T>(fptr, obj);
+    start();
+  }
+  ~thread() { delete func_; }
```
Is std::terminate a thing on the platforms that you're designing this for? Because that is what is supposed to happen if a thread that is still joinable gets destructed. We have seen this arise in real user code (some earlier bug reports that I had to deal with) and it would be good if both versions of the code had the same behavior.

nicolasnoble replied:
Yeah, we have it. I added a check for that.

vjpai commented:
Despite my apparent negativity, I really appreciate the work that has gone into this and feel that this is nearly there. I am also not expecting the grpc:: version to support the full API of the std:: versions of these functions - it is just sufficient to cover the APIs that we use. For that reason, I can see sacrificing some APIs if we don't use them at all. Of my compatibility comments, I would give the most importance to the one on ~thread since that's part of the C++ standard and the proposed change would have very different behavior for threads that get destructed while still  joinable (silent passing in one case, abort in another)

nicolasnoble commented:
Our current usage of unique_lock and thread makes it possible to write the code this way by the way. It is by no mean a true replacement - but can be misleading anyway.

Also, I haven't tried to compile the new version under VS2010 yet.

nicolasnoble commented:
And this compiles fine under Visual Studio 2010.

davidreynolds commented:
Do these new impl/*.h files need to be added to the Makefile? grpc++/server.h includes grpc++/impl/sync.h
